[mlir][tosa] Make TOSA RESIZE's scale, offset, border as Input

Hsiangkai opened the pull request:
Move the `scale`, `offset`, and `border` parameters of the RESIZE operator in the MLIR TOSA dialect from attributes to inputs and update lit tests appropriately.

Add the verifier of the `tosa::ResizeOp` operation.

lhutton1 reviewed (commented):
Thanks @Hsiangkai, LGTM modulo formatting! (I authored part of this so won't explicitly approve)

Jerry-Ge commented:
Thanks @Hsiangkai for the patch. Could we pause merging this before we have a companion TF patch ready to fix RESIZE? 

cc @Tai78641 here as he was the original author who updated this.

Hsiangkai commented:
> Thanks @Hsiangkai for the patch. Could we pause merging this before we have a companion TF patch ready to fix RESIZE?
> 
> cc @Tai78641 here as he was the original author who updated this.

No problem. After you are ready, please help me to merge it.
If there is anything I can help to update this patch, please let me know.

Jerry-Ge commented:
> > Thanks @Hsiangkai for the patch. Could we pause merging this before we have a companion TF patch ready to fix RESIZE?
> > cc @Tai78641 here as he was the original author who updated this.
> 
> No problem. After you are ready, please help me to merge it. If there is anything I can help to update this patch, please let me know.

 Yes, We just got the TF merge access. We will land some previous patches soon (https://github.com/tensorflow/tensorflow/pull/86307). Once we finished those, we then can work on the Resize TF patch. Thanks for being patient!

Jerry-Ge commented:
Hi @Hsiangkai , i have the TF patches ready. Could you rebase this patch then I will take another look to merge it. Thanks for waiting!

Hsiangkai commented:
> Hi @Hsiangkai , i have the TF patches ready. Could you rebase this patch then I will take another look to merge it. Thanks for waiting!

Done. Thanks for your review.

GeorgeARM commented on mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:
```diff
@@ -1444,105 +1447,6 @@ class ResizeUnaryConverter : public OpRewritePattern<tosa::ResizeOp> {
-// TOSA resize with width or height of 1 may be broadcasted to a wider
-// dimension. This is done by materializing a new tosa.resize without
-// the broadcasting behavior, and an explicit broadcast afterwards.
-class MaterializeResizeBroadcast : public OpRewritePattern<tosa::ResizeOp> {
```
Why was this removed? Is this something that could be expanded in the commit message for context?

Hsiangkai replied:
After discussed with the original author, I reverted the rewrite pattern. We can have another patch to remove it later. This patch focuses on TOSA v1.0 update.

GeorgeARM commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -519,6 +526,106 @@ LogicalResult TosaValidation::applyVariableCheck(Operation *op) {
 }
 
+bool checkErrorIfResize(Operation *op) {
+  if (auto resize = dyn_cast<tosa::ResizeOp>(op)) {
```
Can we swap this to early return?

Hsiangkai replied:
Done.

GeorgeARM commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -519,6 +526,106 @@ LogicalResult TosaValidation::applyVariableCheck(Operation *op) {
+
+    const int64_t offsetY = offset[0];
+    const int64_t offsetX = offset[1];
+    const int64_t borderY = border[0];
```
You can move this lower down where you do the border checking

Hsiangkai replied:
Done.

GeorgeARM reviewed (commented):
Minor comments. Thank you @Hsiangkai

Hsiangkai committed:
[mlir][tosa] Make TOSA RESIZE's scale, offset, border as Input

Move the `scale`, `offset`, and `border` parameters of the RESIZE operator
in the MLIR TOSA dialect from attributes to inputs and update lit tests
appropriately.

Add the verifier of the `tosa::ResizeOp` operation.

Co-authored-by: Tai Ly <tai.ly@arm.com>
Co-authored-by: Luke Hutton <luke.hutton@arm.com>

Hsiangkai committed:
Address comments

lhutton1 commented on mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-resize.mlir:
```diff
@@ -38,71 +50,22 @@ func.func @unary_resize_bilinear_fp16(%arg0 : tensor<3x1x1x7xf16>) -> tensor<3x1
 // -----
 
-// CHECK-LABEL: @broadcast_resize_nearest_f32
-func.func @broadcast_resize_nearest_f32(%arg0 : tensor<3x1x1x7xf32>) -> tensor<3x1x5x7xf32> {
```
Since we took the decision to avoid removing `MaterializeResizeBroadcast` for now, we should keep these tests in place. These will likely fail until an exception for this broadcasting is added to https://github.com/llvm/llvm-project/pull/124956/files#diff-90956ba24a2a97cc56a9a3659c7e46e56f1bd791a869246c6a758f9c93f1434fR1762

Hsiangkai replied:
Added back these test cases and check the expected error messages.

lhutton1 replied:
Apologies, there might have been some confusion here. If the intention of keeping `MaterializeResizeBroadcast` was to avoid breaking things for other users that potentially rely on the broadcasting functionality, I think we’d need to make sure the tests that have been replaced don’t have any changes to them (in particular aren’t checking for an expected error). Currently someone that relies on this broadcasting functionality will hit the same error which may break existing flows.

My thinking was to make an exception in the verifier so that it doesn’t check the output shape when broadcasting is involved. We can then repeat this check in the validation pass without the exception to properly check we conform to the specification

Hsiangkai replied:
Thanks for your clarification. It makes sense to me.
I loose the checking in ResizeOp::verify() and add checking to TosaValidation pass.

lhutton1 replied:
Thanks for the updates - this looks good to me now!

Hsiangkai committed:
add back test cases

Hsiangkai committed:
Remove 'materialize broadcast' later

Hsiangkai committed:
Add output size checking to TosaValidation pass

Jerry-Ge reviewed (approved):
LGTM
